crates/matrix-sdk-ui/src/authentication/oidc.rs

```diff
@@ -0,0 +1,283 @@
+    file_path: PathBuf,
+    /// The hash for the metadata used to register the client.
+    /// This is used to check if the client needs to be re-registered.
+    verified_metadata: VerifiedClientMetadata,
+    /// Pre-configured registrations for use with issuers that don't support
+    /// dynamic client registration.
+    static_registrations: HashMap<Url, ClientId>,
+}
+
+/// The underlying data serialized into the registration file.
+#[derive(Debug, Serialize)]
+struct FrozenRegistrationData {
+    /// The hash for the metadata used to register the client.
+    metadata: VerifiedClientMetadata,
+    /// All of the registrations this client has made as a HashMap of issuer URL
+    /// (as a string) to client ID (as a string).
+    dynamic_registrations: HashMap<Url, ClientId>,
+}
+
+/// The deserialize data from the registration file. This data needs to be
+/// validated before it can be used.
+#[derive(Debug, Deserialize)]
+struct UnvalidatedRegistrationData {
+    /// The hash for the metadata used to register the client.
+    metadata: ClientMetadata,
+    /// All of the registrations this client has made as a HashMap of issuer URL
+    /// (as a string) to client ID (as a string).
+    dynamic_registrations: HashMap<Url, ClientId>,
+}
+
+impl UnvalidatedRegistrationData {
+    /// Validates the registration data, returning a `FrozenRegistrationData`.
+    fn validate(&self) -> Result<FrozenRegistrationData, ClientMetadataVerificationError> {
+        let verified_metadata = match self.metadata.clone().validate() {
+            Ok(metadata) => metadata,
+            Err(e) => {
+                tracing::warn!("Failed to validate stored metadata.");
+                return Err(e);
+            }
+        };
+
+        Ok(FrozenRegistrationData {
+            metadata: verified_metadata,
+            dynamic_registrations: self.dynamic_registrations.clone(),
+        })
+    }
+}
+
+/// Manages the storage of OIDC registrations.
+impl OidcRegistrations {
+    /// Creates a new registration store.
+    ///
+    /// # Arguments
+    ///
+    /// * `base_path` - A directory where the registrations file can be stored.
+    ///   It will be nested inside of a directory called `oidc` as
+    ///   `registrations.json`.
+    ///
+    /// * `metadata` - The metadata used to register the client. If this
+    ///   changes, any stored registrations will be lost so the client can
+    ///   re-register with the new data.
+    ///
+    /// * `static_registrations` - Pre-configured registrations for use with
+    ///   issuers that don't support dynamic client registration.
+    pub fn new(
+        base_path: &str,
+        metadata: VerifiedClientMetadata,
+        static_registrations: HashMap<Url, ClientId>,
+    ) -> Result<Self, OidcRegistrationsError> {
+        let oidc_directory = PathBuf::from(base_path).join("oidc");
+        fs::create_dir_all(&oidc_directory).map_err(|_| OidcRegistrationsError::InvalidBasePath)?;
```

Comment:
Rewinding a bit. Why is this using a file on disk, and not one of the stores? :thinking:

Reply:
Because we don't have a store yet. The `AuthenticationService` does everything in memory, only setting up a Client that uses the disk right at the very end of the login request and when the service will be released.

Reply:
Ah right, that makes sense. Thanks.